### ISSUE TITLE
Unable to get good desktop recording quality

### ISSUE THREAD

**Issue by Hubro** [ammen99/wf-recorder#130]
Here is the command I'm running:

```
wf-recorder \
    -x yuv444p \
    -p "preset=faster" \
    -p "crf=14" \
    --audio="alsa_output.pci-0000_11_00.4.analog-stereo.monitor" \
    --file Desktop/test.mp4
```

The log output from the command is:

```
Setting codec option: crf=14
Setting codec option: preset=faster
Setting codec option: tune=zerolatency
Choosing pixel format yuv444p
[libx264 @ 0x7fb4e4002040] using cpu capabilities: MMX2 SSE2Fast SSSE3 SSE4.2 AVX FMA3 BMI2 AVX2
[libx264 @ 0x7fb4e4002040] profile High 4:4:4 Predictive, level 5.2, 4:4:4, 8-bit
[libx264 @ 0x7fb4e4002040] 264 - core 160 r3011 cde9a93 - H.264/MPEG-4 AVC codec - Copyleft 2003-2020 - http://www.videolan.org/x264.html - options: cabac=1 ref=2 deblock=1:0:0 analyse=0x3:0x113 me=hex subme=4 psy=1 psy_rd=1.00:0.00 mixed_ref=0 me_range=16 chroma_me=1 trellis=1 8x8dct=1 cqm=0 deadzone=21,11 fast_pskip=1 chroma_qp_offset=6 threads=22 lookahead_threads=16 sliced_threads=1 slices=22 nr=0 decimate=1 interlaced=0 bluray_compat=0 constrained_intra=0 bframes=0 weightp=1 keyint=250 keyint_min=25 scenecut=40 intra_refresh=0 rc=crf mbtree=0 crf=14.0 qcomp=0.60 qpmin=0 qpmax=69 qpstep=4 ip_ratio=1.40 aq=1:1.00
```

For some reason the video comes out blurry and washed out, here is a comparison:

Screenshot of terminal window:

![real](https://user-images.githubusercontent.com/597206/102812935-3ebc9a80-43c8-11eb-8cb6-706ebf611fcf.png)

Screenshot of recorded video:

![recorded](https://user-images.githubusercontent.com/597206/102812928-3c5a4080-43c8-11eb-9c2a-26d3a4ceeaed.png)

I'm not sure how to debug whether the blurriness is from the encoding or from whatever method wl-recorder is using to capture each frame. From the encoder settings I should be getting as good / better quality than I get from OBS under X11, but I don't. I have tried several different pixel formats, but I can only get a couple of them to work (yuv420p and yuv444p) and they look the same.

**Comment by soreau** [comment]
Sounds like a possible duplicate of #122. wf-recorder uses libx264 by default which encodes using yuv420. Try `-c libx264rgb` to store all the image data.

**Comment by Hubro** [comment]
> Sounds like a possible duplicate of #122. wf-recorder uses libx264 by default which encodes using yuv420. Try `-c libx264rgb` to store all the image data.

Yup, that did it. Thanks!

**Comment by rpetrano** [comment]
I'm afraid the issue is still here and "just use libx264rgb" is not a good answer. It's still impossible to get a decent quality lossy compression video using H.264

Using libx264rgb is only using lossless compression, and takes a ton of space. Also, lossless playback is not supported everywhere.

But using any libx264 encoder seems to produce blurry images. As mentioned in #147 libx264 defaults to using yuv422 which will always result in worse quality, but using yuv444 shouldn't.

However when I try this:
`wf-recorder -l -x yuv444p -p qp=0 -p crf=0 -p preset=veryslow -p tune=animation -c libx264 -f test.mp4`

I get a blurry result. For some reason, if I add `-F format=yuv444p` I get even more blurrier image. That shouldn't happen, as converting from one pixel format to the same shouldn't result in a drastically worse quality.

The same command, but using RGB gives prefect image:
`wf-recorder -l -F format=rgb24 -x rgb24 -p qp=0 -p crf=0 -p preset=veryslow -p tune=animation -c libx264rgb -f testrgb2.mp4`

If I use ffmpeg to convert this RGB24 to yuv444 I also get a perfect result:
`ffmpeg -i testrgb2.mp4 -c:v libx264 -c:a copy -qp 0 -vf format=yuv444p testyuv.mp4`

And the result is much smaller in size, around 10-20% depending on video. If I add compression by removing `-qp 0` and adding `-crf 20` I get very good result, but 2 times smaller size than RGB.

I've tried adding `-F format=yuv444p` to the RGB command above, but it gets even more blurrier than just yuv444 .

I also don't know how to add compression to libx264rgb codec. Seems impossible.

So to summarize, the issue shouldn't be closed because it's still impossible to get a decent quality lossy compression video using H.264

**Comment by formbook** [comment]
yh the rgb solution is absolutely useless because no site will let you upload in that format and the quality just gets destroyed

**Comment by ubitux** [comment]
I don't know who needs to read that, but `qp=0` is the option needed to make a lossless encoding with x264. Unfortunately with wf-recorder this option is not honored, and the reason seems to be because it sets a `crf` option by default. So it is needed to do `-p crf=-1 -p qp=0`. Then setting the pixel format `-x yuv444p` basically gives you a lossless output.
